### PR TITLE
[Service Fabric Managed Clusters] 2021-01-01-preview

### PR DESCRIPTION
**Pull request by a-santamaria**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

We already had a previous APIversion for managed clusters here: [2020-01-01-preview ](https://github.com/Azure/azure-rest-api-specs/tree/master/specification/servicefabric/resource-manager/Microsoft.ServiceFabric/preview/2020-01-01-preview) But we want to move it to its own swagger folder so we can manage APIs and SDKs independently and not face conflicts with models.

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
 Target Date for new api version: 2/8/2021
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  2/5/2021
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/NodeTypeGetOperation_example.json [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    "resourceGroupName": "resRg",
+    "clusterName": "myCluster",
+    "nodeTypeName": "FE",
+    "api-version": "2021-01-01-preview"
+  },
+  "responses": {
+    "200": {
+      "headers": {},
+      "body": {
+        "type": "nodeTypes",
```
Comment (zcxcasd) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/NodeTypeGetOperation_example.json:13]: Microsoft.ServiceFabric/managedClusters/nodeTypes

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterVersionsListByOS_example.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    "resourceGroupName": "resRg",
+    "location": "eastus",
+    "osType": "Windows",
+    "api-version": "2021-01-01-preview"
+  },
+  "responses": {
+    "200": {
+      "headers": {},
+      "body": [
+        {
```
Comment (zcxcasd) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterVersionsListByOS_example.json:13]: doesn't look like a well formatted resource here

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/NodeTypeListOperation_example.json [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    "clusterName": "myCluster",
+    "api-version": "2021-01-01-preview"
+  },
+  "responses": {
+    "200": {
+      "headers": {},
+      "body": {
+        "value": [
+          {
+            "type": "nodeTypes",
```
Comment (zcxcasd) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/NodeTypeListOperation_example.json:14]: fix type to fully qualified

Comment (zcxcasd) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/NodeTypeListOperation_example.json:14]: throughout the file... and other files..

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedcluster.json [hunk 4]
```diff
@@ -0,0 +1,1253 @@
...
+          "description": "The destination address prefix. CIDR or destination IP range. Asterisk '*' can also be used to match all source IPs. Default tags such as 'VirtualNetwork', 'AzureLoadBalancer' and 'Internet' can also be used."
+        },
+        "destinationAddressPrefixes": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "The destination address prefixes. CIDR or destination IP ranges."
+        },
+        "sourcePortRanges": {
```
Comment (pilor) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedcluster.json:878]: having both "sourcePortRange" and "sourcePortRanges" (and similar for the other properties in here) is an artifact of NSGs having to be backwards compatible.  This is a new API that has no baggage so why support both a single property and an array property?  Just have the array.

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json [hunk 5]
```diff
@@ -0,0 +1,2575 @@
...
+          "type": "string",
+          "description": "URL to get the next set of application list results if there are any.",
+          "readOnly": true
+        }
+      }
+    },
+    "ApplicationResourceProperties": {
+      "description": "The application resource properties.",
+      "properties": {
+        "provisioningState": {
```
Comment (pilor) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json:1088]: The async patterns here are a bit all over the place.  The PUTs are attempting to use 202 but then also putting provisioningState in the resource. The correct pattern for async PUT is 201 + provisioning state.  See https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json [hunk 6]
```diff
@@ -0,0 +1,2575 @@
...
+          "readOnly": true,
+          "type": "string",
+          "description": "The current deployment or provisioning state, which only appears in the response"
+        },
+        "typeName": {
```
Comment (pilor) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json:1093]: Is this a reference to an applicationTypes resource?  If so, it should use the full resource ID of the applicationTypes resource (same for typeVersion)

**Review by pilor — COMMENTED** [review]
A few comments.  Up to @ChenTanyi how they feel about splitting out APIs that exist in the service fabric SDK to another SDK

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterVersionsListByOS_example.json (hunk 2 above)
Comment (zcxcasd) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterVersionsListByOS_example.json:13]: missing things like id, type, names, etc.

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedcluster.json (hunk 4 above)
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedcluster.json:878]: @unmeshk12 could you comment on this? are we able to just have the array property?

**Comment by ChenTanyi** [comment]
> A few comments. Up to @ChenTanyi how they feel about splitting out APIs that exist in the service fabric SDK to another SDK

@akning-ms @josefree @PhoenixHe-msft @jianyexi  Would you take a look at it?

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json (hunk 5 above)
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json:1088]: We have the same pattern in all the resources for async operations:
- For new resource, we return 200 with ProvisioningState = Creating
- For existing resources, we return 202 Accepted + Location header, which links to the operation result resource that:
                - returns 202 if the update operation is still running
                - returns 200 if the update operation is completed (resource transition to terminal state)

This is what we discussed with ARM in the initial review a year back when we first introduced these resources, and it looks to be compliant with that doc you shared and the one here https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#put-resource. Are we understanding incorrectly?

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterVersionsListByOS_example.json (hunk 2 above)
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterVersionsListByOS_example.json]: I removed this path. We added it by mistake this is not supported yet and it's going to be different.

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/NodeTypeGetOperation_example.json (hunk 1 above)
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/NodeTypeGetOperation_example.json]: fixed thanks

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedcluster.json (hunk 4 above)
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedcluster.json:702]: thanks, we removed all the non-array properties.

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json (hunk 6 above)
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json]: thanks for the suggestion we changed it to the full resource ID

**Comment by a-santamaria** [comment]
@ChenTanyi I can't figure out what is the issue with the python sdk generation, can you help pelase? 
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=708047&view=logs&j=b7f68bce-9df8-5468-1c9f-74766cfc9a39&t=25fc8ae6-51ed-58e2-b772-f4d054412fe3

**Comment by ChenTanyi** [comment]
> @ChenTanyi I can't figure out what is the issue with the python SDK generation, can you help, please?
> https://dev.azure.com/azure-sdk/internal/_build/results?buildId=708047&view=logs&j=b7f68bce-9df8-5468-1c9f-74766cfc9a39&t=25fc8ae6-51ed-58e2-b772-f4d054412fe3

@msyyc @PhoenixHe-msft Could you take a look at the python generation error?

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterPatchOperation_example.json [hunk 7]
```diff
@@ -0,0 +1,101 @@
+{
+  "parameters": {
+    "subscriptionId": "00000000-0000-0000-0000-000000000000",
+    "resourceGroupName": "resRg",
+    "clusterName": "myCluster",
+    "api-version": "2021-01-01-preview",
+    "parameters": {
+      "tags": {
```
Comment (a-santamaria) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterPatchOperation_example.json:8]: @pilor we currently only support Patch requests to update tags as its required by arm https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#patch-resource. Is it okay to only have the tags in the request, or does it need anything else here like the location?

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json (hunk 5 above)
Comment (pilor) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/managedapplication.json:1246]: Yes, as the doc states provisioningState based updates are preferred but 202 are allowed for back-compat

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterPatchOperation_example.json (hunk 7 above)
Comment (pilor) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview/examples/ManagedClusterPatchOperation_example.json:8]: just tags is fine

**Comment by a-santamaria** [comment]
> > @ChenTanyi I can't figure out what is the issue with the python SDK generation, can you help, please?
> > https://dev.azure.com/azure-sdk/internal/_build/results?buildId=708047&view=logs&j=b7f68bce-9df8-5468-1c9f-74766cfc9a39&t=25fc8ae6-51ed-58e2-b772-f4d054412fe3
> 
> @msyyc @PhoenixHe-msft Could you take a look at the python generation error?

@msyyc @PhoenixHe-msft did you get a chance to look at the issue? I don't think its related to the changes its saying that it can't find bundle:

`09:09:41.198 cmdout 	[automation_generate.sh] INFO:packaging_tools.swaggertosdk.SwaggerToSdkNewCLI:Execute after script: bundle install && rake arm:regen_all_profiles['azure_mgmt_service_fabric_managed_clusters']
09:09:41.777 cmdout 	[automation_generate.sh] INFO:packaging_tools.swaggertosdk.autorest_tools:==[autorest]/bin/sh: 1: bundle: not found
09:09:41.778 cmdout 	[automation_generate.sh] ERROR:packaging_tools.swaggertosdk.autorest_tools:Command 'bundle install && rake arm:regen_all_profiles['azure_mgmt_service_fabric_managed_clusters']' returned non-zero exit status 127.
`

**Comment by a-santamaria** [comment]
@ChenTanyi I fixed the issue with the python sdk generation and it's signed off by ARM. Can we merge the pr?

**Review by ChenTanyi — APPROVED** [review]

**Pull request merged**